File: Python/49_Group_Anagrams.py
class Solution(object):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """

        cach = []
        for i in range(len(strs)):
            st = ''.join(sorted(strs[i]))
            cach.append(st)

        print(cach)
        ht = {}     # store index already searched
        res = []

        for i in range(len(cach)):
            if i not in ht:                
                comb = [strs[i]]
                for j in range(i+1,len(cach)):
                    if cach[i] == cach[j] and j not in ht:
                        comb.append(strs[j])
                        ht[j] = 0
                res.append(comb)

        return res

# character count
# time O(N*M)
# memory O(N*M)

File: C++/49_GroupAnagrams.cpp
using namespace std;
#include <vector>
#include <unordered_map>

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map <string, vector<string>> ht;

        for (auto& str: strs){
            string s = str;
            sort(s.begin(), s.end());
            ht[s].emplace_back(str);
        }

        vector<vector<string>> ans;

        for (auto& h: ht) {
            ans.emplace_back(h.second);
        }

        return ans;
    }
};
